File: go/src/step4_if_fn_do/step4_if_fn_do.go
package main

import (
	"errors"
	"fmt"
	"strings"
)

import (
	"core"
	. "env"
	"printer"
	"reader"
	"readline"
	. "types"
)

// read
func READ(str string) (MalType, error) {
	return reader.Read_str(str)
}

// eval
func eval_ast(ast MalType, env EnvType) (MalType, error) {
	//fmt.Printf("eval_ast: %#v\n", ast)
	if Symbol_Q(ast) {
		return env.Get(ast.(Symbol))
	} else if List_Q(ast) {
		lst := []MalType{}
		for _, a := range ast.(List).Val {
			exp, e := EVAL(a, env)
			if e != nil {
				return nil, e
			}
			lst = append(lst, exp)
		}
		return List{lst, nil}, nil
	} else if Vector_Q(ast) {
		lst := []MalType{}
		for _, a := range ast.(Vector).Val {
			exp, e := EVAL(a, env)
			if e != nil {
				return nil, e
			}
			lst = append(lst, exp)
		}
		return Vector{lst, nil}, nil
	} else if HashMap_Q(ast) {
		m := ast.(HashMap)
		new_hm := HashMap{map[string]MalType{}, nil}
		for k, v := range m.Val {
			ke, e1 := EVAL(k, env)
			if e1 != nil {
				return nil, e1
			}
			if _, ok := ke.(string); !ok {
				return nil, errors.New("non string hash-map key")
			}
			kv, e2 := EVAL(v, env)
			if e2 != nil {
				return nil, e2
			}
			new_hm.Val[ke.(string)] = kv
		}
		return new_hm, nil
	} else {
		return ast, nil
	}
}

func EVAL(ast MalType, env EnvType) (MalType, error) {
	//fmt.Printf("EVAL: %v\n", printer.Pr_str(ast, true))
	switch ast.(type) {
	case List: // continue
	default:
		return eval_ast(ast, env)
	}

	// apply list
	a0 := ast.(List).Val[0]
	var a1 MalType = nil
	var a2 MalType = nil
	switch len(ast.(List).Val) {
	case 1:
		a1 = nil
		a2 = nil
	case 2:
		a1 = ast.(List).Val[1]
		a2 = nil
	default:
		a1 = ast.(List).Val[1]
		a2 = ast.(List).Val[2]
	}
	a0sym := "__<*fn*>__"
	if Symbol_Q(a0) {
		a0sym = a0.(Symbol).Val
	}
	switch a0sym {
	case "def!":
		res, e := EVAL(a2, env)
		if e != nil {
			return nil, e
		}
		return env.Set(a1.(Symbol), res), nil
	case "let*":
		let_env, e := NewEnv(env, nil, nil)
		if e != nil {
			return nil, e
		}
		arr1, e := GetSlice(a1)
		if e != nil {
			return nil, e
		}
		for i := 0; i < len(arr1); i += 2 {
			if !Symbol_Q(arr1[i]) {
				return nil, errors.New("non-symbol bind value")
			}
			exp, e := EVAL(arr1[i+1], let_env)
			if e != nil {
				return nil, e
			}
			let_env.Set(arr1[i].(Symbol), exp)
		}
		return EVAL(a2, let_env)
	case "do":
		el, e := eval_ast(List{ast.(List).Val[1:], nil}, env)
		if e != nil {
			return nil, e
		}
		lst := el.(List).Val
		if len(lst) == 0 {
			return nil, nil
		}
		return lst[len(lst)-1], nil
	case "if":
		cond, e := EVAL(a1, env)
		if e != nil {
			return nil, e
		}
		if cond == nil || cond == false {
			if len(ast.(List).Val) >= 4 {
				return EVAL(ast.(List).Val[3], env)
			} else {
				return nil, nil
			}
		} else {
			return EVAL(a2, env)
		}
	case "fn*":
		return func(arguments []MalType) (MalType, error) {
			new_env, e := NewEnv(env, a1, List{arguments, nil})
			if e != nil {
				return nil, e
			}
			return EVAL(a2, new_env)
		}, nil
	default:
		el, e := eval_ast(ast, env)
		if e != nil {
			return nil, e
		}
		f, ok := el.(List).Val[0].(func([]MalType) (MalType, error))
		if !ok {
			return nil, errors.New("attempt to call non-function")
		}
		return f(el.(List).Val[1:])
	}
}

// print
func PRINT(exp MalType) (string, error) {
	return printer.Pr_str(exp, true), nil
}

var repl_env, _ = NewEnv(nil, nil, nil)

// repl
func rep(str string) (MalType, error) {
	var exp MalType
	var res string
	var e error
	if exp, e = READ(str); e != nil {
		return nil, e
	}
	if exp, e = EVAL(exp, repl_env); e != nil {
		return nil, e
	}
	if res, e = PRINT(exp); e != nil {
		return nil, e
	}
	return res, nil
}

func main() {
	// core.go: defined using go
	for k, v := range core.NS {
		repl_env.Set(Symbol{k}, v)
	}

	// core.mal: defined using the language itself
	rep("(def! not (fn* (a) (if a false true)))")

	// repl loop
	for {
		text, err := readline.Readline("user> ")
		text = strings.TrimRight(text, "\n")
		if err != nil {
			return
		}
		var out MalType
		var e error
		if out, e = rep(text); e != nil {
			if e.Error() == "<empty line>" {
				continue
			}
			fmt.Printf("Error: %v\n", e)
			continue
		}
		fmt.Printf("%v\n", out)
	}
}

File: impls/rpython/step4_if_fn_do.py
import sys, traceback
import mal_readline
import mal_types as types
from mal_types import (MalSym, MalInt, MalStr,
                       nil, true, false, _symbol, _keywordu,
                       throw_str,
                       MalList, _list, MalVector, MalHashMap, MalFunc)
import reader, printer
from env import Env
import core

# read
def READ(str):
    return reader.read_str(str)

# eval
def EVAL(ast, env):
    if env.get(u"DEBUG-EVAL") not in (None, nil, false):
        print(u"EVAL: " + printer._pr_str(ast))
    if types._symbol_Q(ast):
        assert isinstance(ast, MalSym)
        value = env.get(ast.value)
        if value is None:
            throw_str("'" + str(ast.value) + "' not found")
        return value
    elif types._vector_Q(ast):
        res = []
        for a in ast.values:
            res.append(EVAL(a, env))
        return MalVector(res)
    elif types._hash_map_Q(ast):
        new_dct = {}
        for k in ast.dct.keys():
            new_dct[k] = EVAL(ast.dct[k], env)
        return MalHashMap(new_dct)
    elif not types._list_Q(ast):
        return ast  # primitive value, return unchanged
    else:
        # apply list
        if len(ast) == 0: return ast
        a0 = ast[0]
        if isinstance(a0, MalSym):
            a0sym = a0.value
        else:
            a0sym = u"__<*fn*>__"

        if u"def!" == a0sym:
            a1, a2 = ast[1], ast[2]
            res = EVAL(a2, env)
            return env.set(a1, res)
        elif u"let*" == a0sym:
            a1, a2 = ast[1], ast[2]
            let_env = Env(env)
            for i in range(0, len(a1), 2):
                let_env.set(a1[i], EVAL(a1[i+1], let_env))
            return EVAL(a2, let_env)
        elif u"do" == a0sym:
            if len(ast) == 0:
                return nil
            for i in range(1, len(ast) - 1):
                EVAL(ast[i], env)
            return EVAL(ast[-1], env)
        elif u"if" == a0sym:
            a1, a2 = ast[1], ast[2]
            cond = EVAL(a1, env)
            if cond is nil or cond is false:
                if len(ast) > 3: return EVAL(ast[3], env)
                else:            return nil
            else:
                return EVAL(a2, env)
        elif u"fn*" == a0sym:
            a1, a2 = ast[1], ast[2]
            return MalFunc(None, a2, env, a1, EVAL)
        else:
            f = EVAL(a0, env)
            args_list = []
            for i in range(1, len(ast)):
                args_list.append(EVAL(ast[i], env))
            args = MalList(args_list)
            if isinstance(f, MalFunc):
                return f.apply(args)
            else:
                raise Exception("%s is not callable" % f)

# print
def PRINT(exp):
    return printer._pr_str(exp)

# repl
def entry_point(argv):
    repl_env = Env()
    def REP(str, env):
        return PRINT(EVAL(READ(str), env))

    # core.py: defined using python
    for k, v in core.ns.items():
        repl_env.set(_symbol(unicode(k)), MalFunc(v))

    # core.mal: defined using the language itself
    REP("(def! not (fn* (a) (if a false true)))", repl_env)

    while True:
        try:
            line = mal_readline.readline("user> ")
            if line == "": continue
            print(REP(line, repl_env))
        except EOFError as e:
            break
        except reader.Blank:
            continue
        except types.MalException as e:
            print(u"Error: %s" % printer._pr_str(e.object, False))
        except Exception as e:
            print("Error: %s" % e)
            #print("".join(traceback.format_exception(*sys.exc_info())))
    return 0

# _____ Define and setup target ___
def target(*args):
    return entry_point

# Just run entry_point if not RPython compilation
import sys
if not sys.argv[0].endswith('rpython'):
    entry_point(sys.argv)
